File: parallelization_server.R
# install/link packages in renv lockfile
renv::restore()

# load required packages
if (!require("pacman")) install.packages("pacman")
pacman::p_load(
  tidyverse, tictoc, doFuture,
  fixest, broom, furrr, renv
)

# set up parallelization
registerDoFuture()
plan(multiprocess)

# set RNG seed
set.seed(1234321)

# read in data
df <- read_csv("data/final_panel.csv")
df

# Number of bootstrap resamples
N <- 200

# initialize list of times
times <- list()

# initialize storage df
bootstrap_samples <- tibble()

# define bootstrap function
bootstrap_ses <- function(N) {
  df %>%
    dplyr::sample_frac(replace = TRUE) %>%
    fixest::feols(I(log(abundance)) ~ I(log(gdppc)) + temp_avg + precip_avg + interact(year, state) |
                    location_id + year, data = .) %>%
    broom::tidy() %>%
    dplyr::filter(stringr::str_detect(term, "gdppc")) %>%
    dplyr::select(estimate)
}

# bootstrap sequentially
tic()
serial_feols <- replicate(N, bootstrap_ses(N)) %>%
  unlist() %>%
  as_tibble() %>%
  mutate(type = "serial")
time <- toc()
times[["serial_feols"]] <- time$toc - time$tic

# bootstrap in parallel with parallel for loop
tic()
dopar_feols <- foreach(n = seq(N)) %dopar% {
  bootstrap_ses(n)
} %>%
  unlist() %>%
  as_tibble() %>%
  mutate(type = "dopar")
time <- toc()
times[["dopar_feols"]] <- time$toc - time$tic

# bootstrap in parallel with future map
tic()
furrr_feols <- future_map_dfr(
  seq(N),
  bootstrap_ses,
  .options = future_options(globals = "df")
) %>%
  unlist() %>%
  as_tibble() %>%
  mutate(type = "furrr")
time <- toc()
times[["furrr_feols"]] <- time$toc - time$tic

times

bind_rows(serial_feols, dopar_feols, furrr_feols) %>%
  group_by(type) %>%
  summarise(sd(value))

File: parallelization_server.jl
#######################
### PACKAGE LOADING
#######################

# load and install packages on main core
# run(`pip install xlrd`) # install python xlrd package if not on biohpc
using Pkg
Pkg.activate(".")
Pkg.instantiate()
using Distributed, Random, Statistics

# add other workers
addprocs(convert(Int64, length(Sys.cpu_info())/2))

# load and install packages on all cores
@everywhere using Pkg
@everywhere Pkg.activate(".")
@everywhere Pkg.instantiate()
@everywhere using DataFrames, CSV, FixedEffectModels, SharedArrays, Statistics, Random, StatsBase

#######################
### INITIALIZE
#######################

# set RNG seed
Random.seed!(1234321)

# load data
df = CSV.read("data/final_panel.csv", DataFrame)

# set number of bootstraps
N = 200

# preallocate storage arrays
estimates_serial = Vector{Float64}(undef, N);
estimates_distrib = SharedArray{Float64}(N);

# define bootstrap function
@everywhere function bootstrap_ses(df)
    bootstrap_df = df[sample(axes(df, 1), nrow(df); replace = true), :]
    reg_output = reg(bootstrap_df,
        @formula(log(abundance) ~ log(gdppc) + temp_avg + precip_avg + fe(state)&year + fe(location_id) + fe(year)))

    return reg_output.coef[1]
end

#######################
### PRECOMPILE
#######################

# serial
@time for i in 1:N
  estimates_serial[i] = bootstrap_ses(df)
end

# parallel map
@time estimates_pmap = pmap(bootstrap_ses, [df for i in 1:N]);

# distributed
@time @sync @distributed for i in 1:N
  estimates_distrib[i] = bootstrap_ses(df)
end

#######################
### ACTUAL RUN
#######################

@time for i in 1:N
  estimates_serial[i] = bootstrap_ses(df)
end

@time estimates_pmap = pmap(bootstrap_ses, [df for i in 1:N]);

@time @sync @distributed for i in 1:N
  estimates_distrib[i] = bootstrap_ses(df)
end

#######################
### RESULTS
#######################

results = DataFrame(serial = estimates_serial, pmap = estimates_pmap, distrib = estimates_distrib);
std_devs = describe(results, :std)
